api/query_test.go

```diff
@@ -0,0 +1,330 @@
+				http.StatusText(testcase.expectedStatus),
+				testcase.name,
+			)
+		}
+		if testcase.expectedError == "" {
+			parts := readMultipartData(t, w)
+			if len(parts) != 2 {
+				msg := "Got %d parts in reply; want it to always contain 2 in case '%s'"
+				t.Errorf(
+					msg,
+					len(parts),
+					testcase.name,
+				)
+			}
+
+			type SliceAxis struct {
+				Annotation string  `json:"annotation" binding:"required"`
+				Max        float32 `json:"max"        binding:"required"`
+				Min        float32 `json:"min"        binding:"required"`
+				Samples    int     `json:"samples"    binding:"required"`
+				Unit       string  `json:"unit"       binding:"required"`
+			}
+
+			type SliceMetadata struct {
+				Axis   []SliceAxis `json:"axis"   binding:"required"`
+				Format int         `json:"format" binding:"required"`
+			}
+
+			inlineAxis := SliceAxis{
+				Annotation: "Inline", Max: 5.0, Min: 1.0, Samples: 3, Unit: "unitless",
+			}
+			crosslineAxis := SliceAxis{
+				Annotation: "Crossline", Max: 11.0, Min: 10.0, Samples: 2, Unit: "unitless",
+			}
+			sampleAxis := SliceAxis{
+				Annotation: "Sample", Max: 16.0, Min: 4.0, Samples: 4, Unit: "ms",
+			}
+
+			expectedFormat := 3
+
+			var expectedMetadata *SliceMetadata
+
+			switch testcase.slice.Direction {
+			case "i":
+				fallthrough
+			case "inline":
+				expectedMetadata = &SliceMetadata{
+					Axis:   []SliceAxis{crosslineAxis, sampleAxis},
+					Format: expectedFormat}
+			case "j":
+				fallthrough
+			case "crossline":
+				expectedMetadata = &SliceMetadata{
+					Axis:   []SliceAxis{inlineAxis, sampleAxis},
+					Format: expectedFormat}
+			case "k":
+				fallthrough
+			case "depth":
+				fallthrough
+			case "time":
+				fallthrough
+			case "sample":
+				expectedMetadata = &SliceMetadata{
+					Axis:   []SliceAxis{inlineAxis, crosslineAxis},
+					Format: expectedFormat}
+			default:
+				t.Fatalf("Unknown direction %s in case %s", testcase.slice.Direction, testcase.name)
+			}
+
+			metadata := &SliceMetadata{}
+			err := json.Unmarshal(parts[0], metadata)
+
+			if err != nil {
+				msg := "Failed json metadata extraction in case '%s'"
+				t.Fatalf(
+					msg,
+					testcase.name,
+				)
+			}
+
+			if !reflect.DeepEqual(metadata, expectedMetadata) {
+				msg := "Got %v as metadata; want it to be %v in case '%s'"
+				t.Fatalf(
+					msg,
+					metadata,
+					expectedMetadata,
+					testcase.name,
+				)
+			}
+
+			expectedDataLength := expectedMetadata.Axis[0].Samples *
+				expectedMetadata.Axis[1].Samples * 4 //4 bytes each
+			if len(parts[1]) != expectedDataLength {
+				msg := "Got %d bytes in data reply; want it to be %d in case '%s'"
+				t.Errorf(
+					msg,
+					len(parts[2]),
+					expectedDataLength,
+					testcase.name,
+				)
+			}
+		} else {
+			assertError(w, t, struct {
+				name  string
+				error string
+			}{testcase.name, testcase.expectedError})
+		}
+	}
+}
+
+func assertError(w *httptest.ResponseRecorder, t *testing.T,
```

Comment:
isn't it a bit cleaner to have this function return an error, which is asserted in the test? Then you don't need to pass around the test object (`t`).